### PR TITLE
Done #115

### LINKED ISSUE #69: [Feature] [Flag] Request to join
# Feature request: Misc

## Description of feature?

Ability for a user to request to join a group. 
This will notify those who have a paticular GroupFlag when they join that a user wished to join the ranks.

Zones can be set to private (or semi private) where users will not be able to request to join (semi private will exclude only some players from being able to join). 

## Implementation

I am suggesting two flags. 

### Visibility flag

This will change if a player can request to join this group. The status's will be either ``private``, ``semi-private`` and ``public``.
with the other properties of the flag including ``excluded-players``
> Whereby players in this list cannot request to join
and ``include-sub-regions``
> Whereby sub regions cannot be bought nor rented to those who arnt public facing, however if disabled, then despite the zone being private, they can still join by renting/buying a sub region

### LINKED ISSUE #115: [Flag] Fire Damage
# Feature Request: Flag

## Description of flag

>> Prevents damage from fire

## Commands
``/zone flag <zoneId> damage fire <enabled/disabled>``

## Extra info

>> groupkey required
>> set group key by command
>> view flag data

### LINKED ISSUE #157: [Flag] [Feature] Monster Grief
# Feature Request: Flag

## Description of flag

Monster Anti-Grief flag

>> Doesn't let monster grief in zones having this flag

Due to the nature of how events work, this should be broken down into seperate entities. While this could still be stored in a single flag, the prevention for monsters should be seperate

- [x] Creeper (explosion)
- [x] Enderman (taking of block)
- [x] Zombie (knocking down door)
- [x] Skeleton (Setting something on fire)
- [x] Enderdragon (breaking blocks/explosion)
- [x] Wither (breaking blocks/explosion of blocks)
- [x] endermites (hatching/infecting blocks)

### LINKED ISSUE #158: [Flag] [Feature] Farm-Protection
# Feature Request: Flag

## Description of flag

>> Protects farms from trampling

## Commands

>> `/zone region flag <zoneId> farm protection set <enable/disable>`
>> `/zone region flag <zoneId> farm protection view`

### PR DESCRIPTION
**Pull request by Lukamon24**
Closes #115

### CODE REVIEW

File: src/main/java/org/zone/utils/Messages.java [hunk 1]
```diff
@@ -253,7 +255,7 @@ public static Component getZoneFlagLeavingMessageSet(Component message) {
         return Component
                 .text("Leaving message is now: ")
                 .color(NamedTextColor.AQUA)
-                .append(message);
+                .append(message.color(NamedTextColor.GOLD));
```
Comment (mosemister) [inline src/main/java/org/zone/utils/Messages.java]: This should not be forcing a component to be gold. The message should be set to gold before passed in

Comment (Lukamon24) [reply src/main/java/org/zone/utils/Messages.java]: K

Comment (Lukamon24) [reply src/main/java/org/zone/utils/Messages.java]: can't be done bruh

Comment (mosemister) [reply src/main/java/org/zone/utils/Messages.java]: The only place this method is used is on the LeavingSet command. While it comes in as a Component, it is meant to be this as the player can set the colour and the message they are setting in its entirity. So remove this change

File: src/main/java/org/zone/region/flag/meta/join/visibility/flag/ZoneVisibilityFlagType.java [hunk 2]
```diff
@@ -35,18 +35,19 @@
...
+        }
         try {
-            String name = node.node("ZoneVisibility").getString();
-            if (name != null) {
-                ZoneVisibility zoneVisibilityName = ZoneVisibility.valueOf(name);
-                ZoneVisibilityFlag zoneVisibilityFlag = new ZoneVisibilityFlag();
-                zoneVisibilityFlag.setZoneVisibility(zoneVisibilityName);
-                return zoneVisibilityFlag;
-            }
+            ZoneVisibility zoneVisibilityName = ZoneVisibility.valueOf(name);
```
Comment (mosemister) [inline src/main/java/org/zone/region/flag/meta/join/visibility/flag/ZoneVisibilityFlagType.java]: Try something like this

```java
ZoneVisibility visibility;
try{
visibility = ZoneVisibility.valueOf(name);
}catch(IllegalArgumentException e){
    throw new IOException("Invalid name");
}
//rest of your code

```

This way it takes all the try and catch code out of the try and catch making it easier to handle and read

File: src/main/java/org/zone/utils/Messages.java (hunk 1 above)
Comment (Lukamon24) [reply src/main/java/org/zone/utils/Messages.java]: Uhm m

File: src/main/java/org/zone/region/flag/meta/join/visibility/flag/ZoneVisibilityFlagType.java (hunk 2 above)
Comment (Lukamon24) [reply src/main/java/org/zone/region/flag/meta/join/visibility/flag/ZoneVisibilityFlagType.java]: K

Comment (Lukamon24) [reply src/main/java/org/zone/region/flag/meta/join/visibility/flag/ZoneVisibilityFlagType.java]: cant

File: src/main/java/org/zone/region/flag/meta/join/visibility/flag/ZoneVisibilityFlagType.java [hunk 3]
```diff
@@ -35,18 +35,18 @@
...
-            ZoneVisibilityFlag zoneVisibilityFlag = new ZoneVisibilityFlag();
-            zoneVisibilityFlag.setZoneVisibility(zoneVisibilityName);
-            return zoneVisibilityFlag;
+            zoneVisibilityName  = ZoneVisibility.valueOf(name);
         }catch (IllegalArgumentException iae) {
-            iae.printStackTrace();
+            throw new IOException("Invalid Name");
         }
+        ZoneVisibilityFlag zoneVisibilityFlag = new ZoneVisibilityFlag();
+        zoneVisibilityFlag.setZoneVisibility(zoneVisibilityName);
```
Comment (mosemister) [inline src/main/java/org/zone/region/flag/meta/join/visibility/flag/ZoneVisibilityFlagType.java]: You may want to take a look at that last throw. Are you sure you want to throw there?

Comment (Lukamon24) [reply src/main/java/org/zone/region/flag/meta/join/visibility/flag/ZoneVisibilityFlagType.java]: hmm wait

File: src/main/java/org/zone/region/flag/meta/join/visibility/flag/ZoneVisibilityFlagType.java (hunk 2 above)
Comment (Lukamon24) [reply src/main/java/org/zone/region/flag/meta/join/visibility/flag/ZoneVisibilityFlagType.java]: done

File: src/main/java/org/zone/commands/structure/join/invite/InvitePlayerCommand.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        return Optional.empty();
+    }
+
+    @Override
+    public @NotNull CommandResult run(
+            @NotNull CommandContext commandContext, @NotNull String... args) {
+        if (!(commandContext.getSource() instanceof Player player)) {
+            return CommandResult.error(Messages.getPlayerOnlyMessage());
+        }
+        Zone zone = commandContext.getArgument(this, ZONE_ID);
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/invite/InvitePlayerCommand.java]: Seems like this needs work

File: src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilityViewCommand.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        ZoneVisibilityFlag zoneVisibilityFlag = zone
+                .getFlag(FlagTypes.ZONE_VISIBILITY)
+                .orElse(new ZoneVisibilityFlag());
+        String visibilityName = zoneVisibilityFlag.getZoneVisibility().toString();
+        commandContext
+                .getCause()
+                .sendMessage(Identity.nil(), Messages
+                        .getZoneVisibilityTag()
+                        .append(Component.text(visibilityName)
+                                .color(NamedTextColor.GOLD)));
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilityViewCommand.java]: The append part should be done in the .messages class

File: src/main/java/org/zone/commands/structure/region/flags/damage/to/both/tnt/ZoneFlagTntDefuseSetEnableDisableCommand.java [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package org.zone.commands.structure.region.flags.damage.tnt;
+package org.zone.commands.structure.region.flags.damage.to.both.tnt;
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/region/flags/damage/to/both/tnt/ZoneFlagTntDefuseSetEnableDisableCommand.java]: I preferred it the other way around

File: src/main/java/org/zone/commands/structure/region/flags/damage/to/player/attack/ZoneFlagDamageAttackSetEnabledCommand.java [hunk 7]
```diff
@@ -1,4 +1,4 @@
-package org.zone.commands.structure.region.flags.damage.attack;
+package org.zone.commands.structure.region.flags.damage.to.player.attack;
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/region/flags/damage/to/player/attack/ZoneFlagDamageAttackSetEnabledCommand.java]: Not a fan of `to` either revert it or have `towards` or some other word

File: src/main/java/org/zone/commands/structure/join/JoinZoneCommand.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    public @NotNull CommandResult run(
+            @NotNull CommandContext commandContext, @NotNull String... args) {
+        if (!(commandContext.getSource() instanceof Player player)) {
+            return CommandResult.error(Messages.getPlayerOnlyMessage());
+        }
+        Zone zone = commandContext.getArgument(this, ZONE_ID);
+        JoinRequestFlag joinRequestFlag = zone
+                .getFlag(FlagTypes.JOIN_REQUEST)
+                .orElse(new JoinRequestFlag());
+        joinRequestFlag.registerInvite(player.uniqueId());
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/JoinZoneCommand.java]: This is the join class yet your registering a invite? are you sure thats right? if so can we change the wording?

Comment (Lukamon24) [reply src/main/java/org/zone/commands/structure/join/JoinZoneCommand.java]: Umm

File: src/main/java/org/zone/utils/Messages.java [hunk 9]
```diff
@@ -76,7 +87,7 @@ public static Component getNoMessageSet() {
     }
 
     public static Component getFlagMessageView(Component message) {
-        return getMessageTag().append(message);
+        return getMessageTag().append(message.color(NamedTextColor.GOLD));
```
Comment (mosemister) [inline src/main/java/org/zone/utils/Messages.java]: If a component comes into a message, you should not be messing with the format of it. Remove the ``.color(NamedTextColor.GOLD)`` and put it where the component is created

File: src/main/java/org/zone/commands/structure/join/invite/InvitePlayerCommand.java (hunk 4 above)
Comment (Lukamon24) [reply src/main/java/org/zone/commands/structure/join/invite/InvitePlayerCommand.java]: ?

File: src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilityViewCommand.java (hunk 5 above)
Comment (Lukamon24) [reply src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilityViewCommand.java]: K

File: src/main/java/org/zone/commands/structure/region/flags/damage/to/player/attack/ZoneFlagDamageAttackSetEnabledCommand.java (hunk 7 above)
Comment (Lukamon24) [reply src/main/java/org/zone/commands/structure/region/flags/damage/to/player/attack/ZoneFlagDamageAttackSetEnabledCommand.java]: Towards

File: src/main/java/org/zone/utils/Messages.java (hunk 9 above)
Comment (Lukamon24) [reply src/main/java/org/zone/utils/Messages.java]: K

File: src/main/java/org/zone/commands/structure/region/flags/damage/to/both/tnt/ZoneFlagTntDefuseSetEnableDisableCommand.java (hunk 6 above)
Comment (Lukamon24) [reply src/main/java/org/zone/commands/structure/region/flags/damage/to/both/tnt/ZoneFlagTntDefuseSetEnableDisableCommand.java]: Ok

File: src/main/java/org/zone/commands/structure/join/JoinZoneCommand.java (hunk 8 above)
Comment (mosemister) [reply src/main/java/org/zone/commands/structure/join/JoinZoneCommand.java]: Change the method to something like ``registerJoinRequest``

File: src/main/java/org/zone/commands/structure/join/invite/InvitePlayerCommand.java (hunk 4 above)
Comment (mosemister) [reply src/main/java/org/zone/commands/structure/join/invite/InvitePlayerCommand.java]: The method seems unfinished

Comment (Lukamon24) [reply src/main/java/org/zone/commands/structure/join/invite/InvitePlayerCommand.java]: yes

File: src/main/java/org/zone/commands/structure/join/JoinZoneCommand.java (hunk 8 above)
Comment (Lukamon24) [reply src/main/java/org/zone/commands/structure/join/JoinZoneCommand.java]: done but `registerJoin`

File: src/main/java/org/zone/utils/Messages.java (hunk 9 above)
Comment (Lukamon24) [reply src/main/java/org/zone/utils/Messages.java]: can;t bruh

File: src/main/java/org/zone/commands/structure/join/invite/ZoneInvitePlayerCommand.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    @Override
+    public @NotNull CommandResult run(
+            @NotNull CommandContext commandContext, @NotNull String... args) {
+        Zone zone = commandContext.getArgument(this, ZONE_ID);
+        List<GameProfile> players = commandContext.getArgument(this, USERS);
+        JoinRequestFlag joinRequestFlag = zone
+                .getFlag(FlagTypes.JOIN_REQUEST)
+                .orElse(new JoinRequestFlag());
+        joinRequestFlag.registerInvites(players.stream().map(GameProfile::uuid).collect(Collectors.toList()));
+        players
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/invite/ZoneInvitePlayerCommand.java]: I would move this to into the save (to line 73), just so if the saving fails, the player doesnt get the message

File: src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilityViewCommand.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        return Optional.of(ZonePermissions.ZONE_VISIBILITY_VIEW);
+    }
+
+    @Override
+    public @NotNull CommandResult run(
+            @NotNull CommandContext commandContext, @NotNull String... args) {
+        Zone zone = commandContext.getArgument(this, ZONE_ID);
+        ZoneVisibilityFlag zoneVisibilityFlag = zone
+                .getFlag(FlagTypes.ZONE_VISIBILITY)
+                .orElse(new ZoneVisibilityFlag());
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilityViewCommand.java]: We dont want to create a new flag if its not there on the view. 

if the flag isnt there, then assume ZoneVisibility.PUBLIC, so I would do

``zone.getFlag(FlagTypes.ZONE_VISIBILITY).map(flag -> flag.getVisibility()).orElse(ZoneVisibility.PUBLIC);``

File: src/main/java/org/zone/commands/structure/join/invite/ZoneInvitePlayerCommand.java (hunk 10 above)
Comment (Lukamon24) [reply src/main/java/org/zone/commands/structure/join/invite/ZoneInvitePlayerCommand.java]: done

File: src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilityViewCommand.java (hunk 11 above)
Comment (Lukamon24) [reply src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilityViewCommand.java]: done......

File: src/main/java/org/zone/region/flag/entity/player/damage/fire/PlayerFireDamageFlagType.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+import org.jetbrains.annotations.NotNull;
+import org.spongepowered.plugin.PluginContainer;
+import org.zone.ZonePlugin;
+import org.zone.region.flag.FlagType;
+
+public class PlayerFireDamageFlagType implements FlagType.TaggedFlagType<PlayerFireDamageFlag> {
+
+    String NAME = "Prevent Fire Damage To Players";
+    PluginContainer CONTAINER = ZonePlugin.getZonesPlugin().getPluginContainer();
```
Comment (mosemister) [inline src/main/java/org/zone/region/flag/entity/player/damage/fire/PlayerFireDamageFlagType.java]: This is breaking the tests, doesnt need to be stored

File: src/main/java/org/zone/utils/Messages.java (hunk 9 above)
Comment (mosemister) [reply src/main/java/org/zone/utils/Messages.java]: Remove it then

File: src/main/java/org/zone/utils/Messages.java (hunk 1 above)
Comment (mosemister) [reply src/main/java/org/zone/utils/Messages.java]: Still waiting for removal

File: src/main/java/org/zone/region/flag/entity/player/damage/fire/PlayerFireDamageFlagType.java (hunk 12 above)
Comment (Lukamon24) [reply src/main/java/org/zone/region/flag/entity/player/damage/fire/PlayerFireDamageFlagType.java]: ah shoot

File: src/main/java/org/zone/utils/Messages.java (hunk 9 above)
Comment (Lukamon24) [reply src/main/java/org/zone/utils/Messages.java]: ok

File: src/main/java/org/zone/commands/structure/join/invite/ZoneInviteAcceptCommand.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
+package org.zone.commands.structure.join.invite;
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/invite/ZoneInviteAcceptCommand.java]: This class should not be in the package of ``join``

File: src/main/java/org/zone/commands/structure/join/invite/ZoneInviteDenyCommand.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
+package org.zone.commands.structure.join.invite;
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/invite/ZoneInviteDenyCommand.java]: This should not be in the package of ``join``

File: src/main/java/org/zone/commands/structure/join/invite/ZoneInvitePlayerCommand.java [hunk 15]
```diff
@@ -0,0 +1,84 @@
+package org.zone.commands.structure.join.invite;
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/invite/ZoneInvitePlayerCommand.java]: This should not be in the package of ``join``

File: src/main/java/org/zone/commands/structure/join/invite/ZoneInvitePlayerViewCommand.java [hunk 16]
```diff
@@ -0,0 +1,72 @@
+package org.zone.commands.structure.join.invite;
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/invite/ZoneInvitePlayerViewCommand.java]: This should not be in a package called ``join``

File: src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilitySetCommand.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
+package org.zone.commands.structure.join.visibility;
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilitySetCommand.java]: This should be in the ``region`` package as thats what your arguments suggest

File: src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilityViewCommand.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
+package org.zone.commands.structure.join.visibility;
```
Comment (mosemister) [inline src/main/java/org/zone/commands/structure/join/visibility/ZoneVisibilityViewCommand.java]: This should be in the ``region`` package as thats what your arguments suggest

File: src/main/java/org/zone/utils/Messages.java (hunk 1 above)
Comment (Lukamon24) [reply src/main/java/org/zone/utils/Messages.java]: already done

**Review by mosemister — APPROVED** [review]
Approved

**Pull request merged**